### PR TITLE
[US039, US040] Created Vector2 frontend module, created specs, refactored backend tests

### PR DESCRIPTION
**Pull request by tomejzen**

### CODE REVIEW

File: GameServer/wwwroot/js/Physics/Vector2.js [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+        if (typeof that._length === "undefined")
+            that._length = Math.sqrt(that.LengthSquared());
+
+        return that._length;
+    };
+
+    that.Normalize = function () {
+
+        if (that.IsDegenerated)
```
Comment (krzysiekprzekwas) [inline GameServer/wwwroot/js/Physics/Vector2.js]: Once again - function call

File: GameServer/wwwroot/js/Physics/Vector2.js [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+        const vector = Vector2.Copy(that);
+        vector.Divide(that.Length());
+        return vector;
+    };
+
+    that.SafeNormalize = function () {
+
+        try {
+            that.Normalize();
```
Comment (krzysiekprzekwas) [inline GameServer/wwwroot/js/Physics/Vector2.js]: Is there a reason for separate safe and unsafe function?

File: GameTests/Vector2Tests.cs [hunk 3]
```diff
@@ -164,45 +164,45 @@ public void ShouldAllowToCalculateSquaredDistanceFrom()
...
-            var a = new Vector2(-1, 1);
-            var copy = new Vector2(a);
+            var vector = new Vector2(-1, 1);
+            var copy = new Vector2(vector);
 
             // Act
-            var b = a.Normalize();
-            var c = Vector2.Normalize(a);
+            var normalized = vector.Normalize();
+            var staticalyNormalized = Vector2.Normalize(vector);
```
Comment (krzysiekprzekwas) [inline GameTests/Vector2Tests.cs]: What does it check?

File: GameServer/wwwroot/js/Physics/Vector2.js (hunk 2 above)
Comment (tomejzen) [reply GameServer/wwwroot/js/Physics/Vector2.js]: SafeNormalize function is used at Physics when as deafult we want to get ZERO_VECTOR.

File: GameTests/Vector2Tests.cs (hunk 3 above)
Comment (tomejzen) [reply GameTests/Vector2Tests.cs]: It checks normalize method for vector instace as also static method of Vector2 class.

File: GameServer/wwwroot/js/Physics/Vector2.js (hunk 1 above)
Comment (tomejzen) [reply GameServer/wwwroot/js/Physics/Vector2.js]: Fixed

File: GameTests/Vector2Tests.cs (hunk 3 above)
Comment (krzysiekprzekwas) [reply GameTests/Vector2Tests.cs]: Isn't it the same method? I'm not so familiar with JS testing. If this is helpfull lets leave it

File: GameServer/wwwroot/js/Physics/Vector2.js (hunk 2 above)
Comment (krzysiekprzekwas) [reply GameServer/wwwroot/js/Physics/Vector2.js]: How about standard Normalize, is it used anywhere? Why not return new vector after IsDegenereted check?

**Comment by krzysiekprzekwas** [comment]
Amazing

**Pull request merged**